### PR TITLE
Improved TLS handling and add support for ALPN

### PR DESCRIPTION
**Pull request by anubhav7nov**

### CODE REVIEW

**Comment by anubhav7nov** [comment]
# #42

File: paho/build.gradle.kts [hunk 1]
```diff
@@ -11,18 +14,26 @@ ext {
 }
 
 plugins {
-    id("java-library")
-    kotlin("jvm")
+    id("com.android.library")
+    kotlin("android")
+    kotlin("android.extensions")
+    kotlin("kapt")
+    //kotlin("jvm")
```
Comment (deepanshu42) [inline paho/build.gradle.kts:21]: remove commented code

File: paho/build.gradle.kts [hunk 2]
```diff
@@ -11,18 +14,26 @@ ext {
...
 }
 
-java {
-    sourceCompatibility = JavaVersion.VERSION_1_8
-    targetCompatibility = JavaVersion.VERSION_1_8
-}
+//java {
+//    sourceCompatibility = JavaVersion.VERSION_1_8
+//    targetCompatibility = JavaVersion.VERSION_1_8
+//}
```
Comment (deepanshu42) [inline paho/build.gradle.kts:27]: remove commented code

File: paho/src/main/java/org/eclipse/paho/client/mqttv3/internal/platform/android/AndroidLog.kt [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+//        if (configuredLoggers.add(logger)) {
+//            logger.useParentHandlers = false
+//            // log based on levels at startup to avoid logging each frame
+//            logger.level = when {
+//                Log.isLoggable(tag, Log.DEBUG) -> Level.FINE
+//                Log.isLoggable(tag, Log.INFO) -> Level.INFO
+//                else -> Level.WARNING
+//            }
+//            logger.addHandler(AndroidLogHandler)
+    }
```
Comment (deepanshu42) [inline paho/src/main/java/org/eclipse/paho/client/mqttv3/internal/platform/android/AndroidLog.kt:112]: can be removed

File: paho/src/main/java/org/eclipse/paho/client/mqttv3/internal/platform/android/AndroidLog.kt [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+//        val packageName = OkHttpClient::class.java.`package`?.name
+//
+//        if (packageName != null) {
+//            this[packageName] = "OkHttp"
+//        }
+//
+//        this[OkHttpClient::class.java.name] = "okhttp.OkHttpClient"
+//        this[Http2::class.java.name] = "okhttp.Http2"
+//        this[TaskRunner::class.java.name] = "okhttp.TaskRunner"
+//        this["okhttp3.mockwebserver.MockWebServer"] = "okhttp.MockWebServer"
```
Comment (deepanshu42) [inline paho/src/main/java/org/eclipse/paho/client/mqttv3/internal/platform/android/AndroidLog.kt:63]: remove commented code

File: mqtt-client/src/main/java/com/gojek/mqtt/model/MqttConnectOptions.kt [hunk 5]
```diff
@@ -1,18 +1,216 @@
...
+            this.sslSocketFactoryOrNull = sslSocketFactory
+            this.x509TrustManagerOrNull = trustManager
+        }
+
+        fun connectionSpec(connectionSpec: ConnectionSpec) = apply {
+            this.connectionSpec = connectionSpec
+        }
+
+        fun alpnProtocols(protocols: List<Protocol>) = apply {
+            require(protocols.isNotEmpty()) { "password cannot be empty" }
```
Comment (deepanshu42) [inline mqtt-client/src/main/java/com/gojek/mqtt/model/MqttConnectOptions.kt:200]: incorrect error message

File: mqtt-client/src/main/java/com/gojek/mqtt/model/MqttConnectOptions.kt [hunk 6]
```diff
@@ -1,18 +1,216 @@
...
+        fun password(password: String) = apply {
+            this.password = password
+        }
+
+        fun cleanSession(cleanSession: Boolean) = apply {
+            this.isCleanSession = cleanSession
+        }
+
+        fun readTimeoutSecs(readTimeoutSecs: Int) = apply {
+            require(readTimeoutSecs > 0 || readTimeoutSecs == DEFAULT_READ_TIMEOUT) { "read timeout should be >= $DEFAULT_READ_TIMEOUT" }
```
Comment (deepanshu42) [inline mqtt-client/src/main/java/com/gojek/mqtt/model/MqttConnectOptions.kt:123]: can we simply make it `require(readTimeoutSecs > 0)`?

File: mqtt-client/src/main/java/com/gojek/mqtt/model/MqttConnectOptions.kt [hunk 7]
```diff
@@ -1,18 +1,216 @@
...
+            this.x509TrustManager = null
+        } else if (builder.sslSocketFactoryOrNull != null) {
+            this.sslSocketFactory = builder.sslSocketFactoryOrNull
+            this.x509TrustManager = builder.x509TrustManagerOrNull!!
+        } else {
+            this.x509TrustManager = Platform.get().platformTrustManager()
+            this.sslSocketFactory = Platform.get().newSslSocketFactory(x509TrustManager!!)
+        }
+
+        this.readTimeoutSecs = if (builder.readTimeoutSecs == DEFAULT_READ_TIMEOUT) {
```
Comment (deepanshu42) [inline mqtt-client/src/main/java/com/gojek/mqtt/model/MqttConnectOptions.kt:56]: if (builder.readTimeoutSecs < builder.keepAlive.timeSeconds)

File: mqtt-client/src/main/java/com/gojek/mqtt/model/KeepAlive.kt [hunk 8]
```diff
@@ -3,4 +3,8 @@ package com.gojek.mqtt.model
 data class KeepAlive(
     val timeSeconds: Int,
     internal val isOptimal: Boolean = false
-)
+) {
+    companion object {
+        val NO_KEEP_ALIVE = KeepAlive(timeSeconds = -1, isOptimal = false)
```
Comment (deepanshu42) [inline mqtt-client/src/main/java/com/gojek/mqtt/model/KeepAlive.kt:8]: can we use a positive value like 60 seconds for default keepalive?

File: paho/build.gradle.kts (hunk 2 above)
Comment (anubhav7nov) [reply paho/build.gradle.kts]: done

File: paho/src/main/java/org/eclipse/paho/client/mqttv3/internal/platform/android/AndroidLog.kt (hunk 4 above)
Comment (anubhav7nov) [reply paho/src/main/java/org/eclipse/paho/client/mqttv3/internal/platform/android/AndroidLog.kt]: done

File: mqtt-client/src/main/java/com/gojek/mqtt/model/MqttConnectOptions.kt (hunk 6 above)
Comment (anubhav7nov) [reply mqtt-client/src/main/java/com/gojek/mqtt/model/MqttConnectOptions.kt]: done

File: mqtt-client/src/main/java/com/gojek/mqtt/model/MqttConnectOptions.kt (hunk 7 above)
Comment (anubhav7nov) [reply mqtt-client/src/main/java/com/gojek/mqtt/model/MqttConnectOptions.kt]: done

File: mqtt-client/src/main/java/com/gojek/mqtt/model/KeepAlive.kt (hunk 8 above)
Comment (anubhav7nov) [reply mqtt-client/src/main/java/com/gojek/mqtt/model/KeepAlive.kt]: done

**Review by deepanshu42 — APPROVED** [review]

**Pull request merged**